File: mf2/javascript/src/relative_time_core.js
import { selectCardinal } from "./cldr_plural_rules.js";
import { isDecimalSourceFunction } from "./function_support.js";
import { localeLookupChain } from "./locale-key.js";

const DEFAULT_LOCALE = "en";
const MAX_LOCALE_LENGTH = 256;
const MAX_OPTION_LENGTH = 256;
const MAX_OPERAND_LENGTH = 256;
const MAX_RELATIVE_TIME_QUANTITY = 1_000_000_000;
const DECIMAL_NUMBER_RE = /^-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?$/;
const STYLE_VALUES = ["long", "short", "narrow"];
const NUMERIC_VALUES = ["always", "auto"];
const POLICY_VALUES = ["precise", "compact", "chat"];
const UNIT_SECONDS = Object.freeze({
  second: 1,
  minute: 60,
  hour: 3_600,
  day: 86_400,
  week: 604_800,
  month: 2_592_000,
  quarter: 7_776_000,
  year: 31_536_000,
});
const UNIT_VALUES = ["auto", ...Object.keys(UNIT_SECONDS)];
const POLICIES = Object.freeze({
  precise: Object.freeze([
    [60, "second"],
    [3_600, "minute"],
    [86_400, "hour"],
    [604_800, "day"],
    [2_592_000, "week"],
    [31_536_000, "month"],
    [Number.POSITIVE_INFINITY, "year"],
  ]),
  compact: Object.freeze([
    [60, "second"],
    [3_600, "minute"],
    [86_400, "hour"],
    [Number.POSITIVE_INFINITY, "day"],
  ]),
  chat: Object.freeze([
    [45, "second"],
    [2_700, "minute"],
    [79_200, "hour"],
    [604_800, "day"],
    [Number.POSITIVE_INFINITY, "week"],
  ]),
});
const DATA_CACHE = new WeakMap();

export class RelativeTimeCoreError extends Error {
  constructor(code, message) {
    super(message);
    this.name = "RelativeTimeCoreError";
    this.code = code;
  }
}

export function formatRelativeTimeCore(value, options = {}) {
  const data = preparedData(options.data);
  const locale = localeOption(options.locale);
  const style = optionOneOf(options.style ?? "short", STYLE_VALUES, "style");
  const numeric = optionOneOf(options.numeric ?? "always", NUMERIC_VALUES, "numeric");
  const policy = optionOneOf(options.policy ?? "precise", POLICY_VALUES, "policy");
  const unit = optionOneOf(options.unit ?? "auto", UNIT_VALUES, "unit");
  const seconds = parseFiniteNumber(value);
  const selectedUnit = unit === "auto" ? selectUnit(seconds, policy) : unit;
  const quantity = relativeTimeQuantity(seconds, selectedUnit);

  if (useRelativeZero(policy, numeric, seconds)) {
    const relative = relativeTerm(data, locale, style, selectedUnit, "0");
    if (relative != null) return relative;
  }
  if (numeric === "auto") {
    const offset = relativeOffset(seconds, selectedUnit, quantity);
    if (offset != null) {
      const relative = relativeTerm(data, locale, style, selectedUnit, offset);
      if (relative != null) return relative;
    }
  }

  const direction = isNegativeRelativeTime(seconds) ? "past" : "future";
  const category = selectCardinal(locale, quantity);
  const pattern = relativeTimePattern(data, locale, style, selectedUnit, direction, category);
  return pattern.replace("{0}", String(quantity));
}

export function formatRelativeTimeCoreToParts(value, options = {}) {
  return [{ type: "text", value: formatRelativeTimeCore(value, options) }];
}

export function createRelativeTimeCoreFunctionRegistry(FunctionRegistry, data) {
  preparedData(data);
  return FunctionRegistry.portable().withFunction("relativeTime", (call) => {
    const options = {
      locale: call.locale,
      style: call.optionValue("style", "short"),
      numeric: call.optionValue("numeric", "always"),
      policy: call.optionValue("policy", "precise"),
      unit: call.optionValue("unit", "auto"),
      data,
    };
    try {
      return formatRelativeTimeCore(call.rawValue ?? call.value, options);
    } catch (error) {
      const sourceValue = relativeTimeSourceValue(call.inheritedSource);
      if (!(error instanceof RelativeTimeCoreError) || error.code !== "bad-operand" || sourceValue == null) {
        throw error;
      }
      return formatRelativeTimeCore(sourceValue, options);
    }
  });
}

function relativeTimeSourceValue(source) {
  for (let current = source; current != null; current = current.inherited) {
    if (isDecimalSourceFunction(current.function)) return current.value;
  }
  return null;
}

function preparedData(data) {
  if (data == null || typeof data !== "object") {
    throw new RelativeTimeCoreError("missing-locale-data", "Relative-time core requires generated CLDR data.");
  }
  const cached = DATA_CACHE.get(data);
  if (cached != null) return cached;
  const localeMap = preparedLocaleMap(data.localeMap);
  if (localeMap == null || !Array.isArray(data.patternSets)) {
    throw new RelativeTimeCoreError("missing-locale-data", "Relative-time core data has an unsupported shape.");
  }
  const patternSets = new Map();
  for (const item of data.patternSets) {
    if (
      item != null &&
      typeof item.id === "string" &&
      item.id !== "" &&
      item.data != null &&
      typeof item.data === "object" &&
      !Array.isArray(item.data) &&
      Object.keys(item.data).length > 0
    ) {
      patternSets.set(item.id, item.data);
    }
  }
  if (patternSets.size === 0) {
    throw new RelativeTimeCoreError("missing-locale-data", "Relative-time core data has an unsupported shape.");
  }
  const prepared = { localeMap, patternSets };
  DATA_CACHE.set(data, prepared);
  return prepared;
}

function preparedLocaleMap(localeMap) {
  if (localeMap == null || typeof localeMap !== "object" || Array.isArray(localeMap)) {
    return null;
  }
  const entries = Object.entries(localeMap);
  if (entries.length === 0 || entries.some(([, setId]) => typeof setId !== "string")) {
    return null;
  }
  return Object.fromEntries(entries);
}

function parseFiniteNumber(value) {
  if (typeof value === "boolean") {
    throw new RelativeTimeCoreError("bad-operand", "Relative-time core requires a finite numeric value.");
  }
  const text = coerceStringOperand(value ?? "").trim();
  if (text === "") {
    throw new RelativeTimeCoreError("bad-operand", "Relative-time core requires a finite numeric value.");
  }
  if (text.length > MAX_OPERAND_LENGTH) {
    throw new RelativeTimeCoreError("bad-operand", "Relative-time core requires a finite numeric value.");
  }
  if (typeof value !== "number" && !DECIMAL_NUMBER_RE.test(text)) {
    throw new RelativeTimeCoreError("bad-operand", "Relative-time core requires a finite numeric value.");
  }
  const parsed = typeof value === "number" ? value : Number(text);
  if (!Number.isFinite(parsed)) {
    throw new RelativeTimeCoreError("bad-operand", "Relative-time core requires a finite numeric value.");
  }
  return parsed;
}

function localeOption(value) {
  if (value == null) return DEFAULT_LOCALE;
  let locale;
  try {
    locale = String(value);
  } catch {
    throw new RelativeTimeCoreError("bad-option", "locale must be coercible to a string.");
  }
  if (locale.length > MAX_LOCALE_LENGTH) {
    throw new RelativeTimeCoreError("bad-option", "locale must not exceed 256 characters.");
  }
  return locale === "" ? DEFAULT_LOCALE : locale;
}

function coerceStringOperand(value) {
  try {
    return String(value);
  } catch {
    throw new RelativeTimeCoreError("bad-operand", "Relative-time core requires a finite numeric value.");
  }
}

function optionOneOf(value, allowed, name) {
  let text;
  try {
    text = String(value);
  } catch {
    throw new RelativeTimeCoreError("bad-option", `${name} must be coercible to a string.`);
  }
  if (text.length > MAX_OPTION_LENGTH) {
    throw new RelativeTimeCoreError("bad-option", `${name} must not exceed 256 characters.`);
  }
  if (!allowed.includes(text)) {
    throw new RelativeTimeCoreError("bad-option", `${name} must be one of ${allowed.join(", ")}.`);
  }
  return text;
}

function selectUnit(seconds, policy) {
  const absolute = Math.abs(seconds);
  for (const [upper, unit] of POLICIES[policy]) {
    if (absolute < upper) return unit;
  }
  return "year";
}

function relativeTimeQuantity(seconds, unit) {
  const absolute = Math.abs(seconds);
  if (absolute === 0) return 0;
  const quantity = Math.max(1, Math.floor(absolute / UNIT_SECONDS[unit] + 0.5));
  if (quantity > MAX_RELATIVE_TIME_QUANTITY) {
    throw new RelativeTimeCoreError("bad-operand", "Relative-time core quantity is outside the supported range.");
  }
  return quantity;
}

function useRelativeZero(policy, numeric, seconds) {
  return policy === "chat" && numeric === "auto" && Math.abs(seconds) < 45;
}

function relativeOffset(seconds, unit, quantity) {
  if (quantity === 0) return "0";
  if (Math.abs(seconds) !== quantity * UNIT_SECONDS[unit]) return null;
  return isNegativeRelativeTime(seconds) ? `-${quantity}` : String(quantity);
}

function isNegativeRelativeTime(seconds) {
  return seconds < 0 || Object.is(seconds, -0);
}

function relativeTerm(data, locale, style, unit, offset) {
  return relativeUnitData(data, locale, style, unit).relative?.[offset] ?? null;
}

function relativeTimePattern(data, locale, style, unit, direction, category) {
  const patterns = relativeUnitData(data, locale, style, unit)[direction];
  const pattern = patterns?.[category] ?? patterns?.other;
  if (typeof pattern === "string") return pattern;
  throw new RelativeTimeCoreError(
    "missing-locale-data",
    `Missing relative-time pattern for ${locale}/${style}/${unit}/${direction}.`,
  );
}

function relativeUnitData(data, locale, style, unit) {
  const patternSet = patternSetFor(data, locale);
  const unitData = patternSet?.[style]?.[unit];
  if (unitData != null) return unitData;
  throw new RelativeTimeCoreError(
    "missing-locale-data",
    `Missing relative-time unit data for ${locale}/${style}/${unit}.`,
  );
}

function patternSetFor(data, locale) {
  for (const candidate of localeLookupChain(locale)) {
    const setId = data.localeMap[candidate];
    if (setId == null) continue;
    const patternSet = data.patternSets.get(setId);
    if (patternSet != null) return patternSet;
    throw new RelativeTimeCoreError("missing-locale-data", `Missing relative-time pattern set ${setId}.`);
  }
  throw new RelativeTimeCoreError("missing-locale-data", `Missing relative-time locale data for ${locale}.`);
}

File: mf2/go/relative_time_core.go
package mf2

import (
	"fmt"
	"math"
	"regexp"
	"strconv"
	"strings"
)

const defaultRelativeTimeCoreLocale = "en"
const maxRelativeTimeCoreOptionLength = 256
const maxRelativeTimeCoreOperandLength = 256
const maxRelativeTimeCoreQuantity = 1_000_000_000

var relativeTimeCoreDecimalNumberPattern = regexp.MustCompile(`^-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?$`)

const (
	RelativeTimeCoreStyleLong   = "long"
	RelativeTimeCoreStyleShort  = "short"
	RelativeTimeCoreStyleNarrow = "narrow"
)

const (
	RelativeTimeCoreNumericAlways = "always"
	RelativeTimeCoreNumericAuto   = "auto"
)

const (
	RelativeTimeCorePolicyPrecise = "precise"
	RelativeTimeCorePolicyCompact = "compact"
	RelativeTimeCorePolicyChat    = "chat"
)

const (
	RelativeTimeCoreUnitAuto    = "auto"
	RelativeTimeCoreUnitSecond  = "second"
	RelativeTimeCoreUnitMinute  = "minute"
	RelativeTimeCoreUnitHour    = "hour"
	RelativeTimeCoreUnitDay     = "day"
	RelativeTimeCoreUnitWeek    = "week"
	RelativeTimeCoreUnitMonth   = "month"
	RelativeTimeCoreUnitQuarter = "quarter"
	RelativeTimeCoreUnitYear    = "year"
)

type RelativeTimeCoreOptions struct {
	Locale  string
	Style   string
	Numeric string
	Policy  string
	Unit    string
}

type RelativeTimeCoreFormatter struct {
	data relativeTimeCorePreparedData
}

type RelativeTimeCoreData struct {
	LocaleMap   map[string]string            `json:"localeMap"`
	PatternSets []RelativeTimeCorePatternSet `json:"patternSets"`
}

type RelativeTimeCorePatternSet struct {
	ID   string                                         `json:"id"`
	Data map[string]map[string]RelativeTimeCoreUnitData `json:"data"`
}

type RelativeTimeCoreUnitData struct {
	Future   map[string]string `json:"future"`
	Past     map[string]string `json:"past"`
	Relative map[string]string `json:"relative"`
}

type relativeTimeCorePreparedData struct {
	localeMap   map[string]string
	patternSets map[string]map[string]map[string]RelativeTimeCoreUnitData
}

type relativeTimeCorePolicyStep struct {
	upper float64
	unit  string
}

var relativeTimeCoreStyleValues = []string{
	RelativeTimeCoreStyleLong,
	RelativeTimeCoreStyleShort,
	RelativeTimeCoreStyleNarrow,
}

var relativeTimeCoreNumericValues = []string{
	RelativeTimeCoreNumericAlways,
	RelativeTimeCoreNumericAuto,
}

var relativeTimeCorePolicyValues = []string{
	RelativeTimeCorePolicyPrecise,
	RelativeTimeCorePolicyCompact,
	RelativeTimeCorePolicyChat,
}

var relativeTimeCoreUnitValues = []string{
	RelativeTimeCoreUnitAuto,
	RelativeTimeCoreUnitSecond,
	RelativeTimeCoreUnitMinute,
	RelativeTimeCoreUnitHour,
	RelativeTimeCoreUnitDay,
	RelativeTimeCoreUnitWeek,
	RelativeTimeCoreUnitMonth,
	RelativeTimeCoreUnitQuarter,
	RelativeTimeCoreUnitYear,
}

var relativeTimeCoreUnitSeconds = map[string]float64{
	RelativeTimeCoreUnitSecond:  1,
	RelativeTimeCoreUnitMinute:  60,
	RelativeTimeCoreUnitHour:    3600,
	RelativeTimeCoreUnitDay:     86400,
	RelativeTimeCoreUnitWeek:    604800,
	RelativeTimeCoreUnitMonth:   2592000,
	RelativeTimeCoreUnitQuarter: 7776000,
	RelativeTimeCoreUnitYear:    31536000,
}

var relativeTimeCorePolicies = map[string][]relativeTimeCorePolicyStep{
	RelativeTimeCorePolicyPrecise: {
		{upper: 60, unit: RelativeTimeCoreUnitSecond},
		{upper: 3600, unit: RelativeTimeCoreUnitMinute},
		{upper: 86400, unit: RelativeTimeCoreUnitHour},
		{upper: 604800, unit: RelativeTimeCoreUnitDay},
		{upper: 2592000, unit: RelativeTimeCoreUnitWeek},
		{upper: 31536000, unit: RelativeTimeCoreUnitMonth},
		{upper: math.Inf(1), unit: RelativeTimeCoreUnitYear},
	},
	RelativeTimeCorePolicyCompact: {
		{upper: 60, unit: RelativeTimeCoreUnitSecond},
		{upper: 3600, unit: RelativeTimeCoreUnitMinute},
		{upper: 86400, unit: RelativeTimeCoreUnitHour},
		{upper: math.Inf(1), unit: RelativeTimeCoreUnitDay},
	},
	RelativeTimeCorePolicyChat: {
		{upper: 45, unit: RelativeTimeCoreUnitSecond},
		{upper: 2700, unit: RelativeTimeCoreUnitMinute},
		{upper: 79200, unit: RelativeTimeCoreUnitHour},
		{upper: 604800, unit: RelativeTimeCoreUnitDay},
		{upper: math.Inf(1), unit: RelativeTimeCoreUnitWeek},
	},
}

func RelativeTimeCoreFunctionRegistry(data RelativeTimeCoreData) FunctionRegistry {
	formatter, prepareErr := NewRelativeTimeCoreFormatter(data)
	if prepareErr == nil {
		return formatter.FunctionRegistry()
	}
	return PortableFunctionRegistry().WithFunction("relativeTime", func(call FunctionCall) (string, error) {
		return "", prepareErr
	})
}

func NewRelativeTimeCoreFormatter(data RelativeTimeCoreData) (RelativeTimeCoreFormatter, error) {
	prepared, err := prepareRelativeTimeCoreData(data)
	if err != nil {
		return RelativeTimeCoreFormatter{}, err
	}
	return RelativeTimeCoreFormatter{data: prepared}, nil
}

func FormatRelativeTimeCore(value any, data RelativeTimeCoreData, options RelativeTimeCoreOptions) (string, error) {
	formatter, err := NewRelativeTimeCoreFormatter(data)
	if err != nil {
		return "", err
	}
	return formatter.Format(value, options)
}

func FormatRelativeTimeCoreToParts(value any, data RelativeTimeCoreData, options RelativeTimeCoreOptions) ([]Part, error) {
	formatter, err := NewRelativeTimeCoreFormatter(data)
	if err != nil {
		return nil, err
	}
	return formatter.FormatToParts(value, options)
}

func (formatter RelativeTimeCoreFormatter) Format(value any, options RelativeTimeCoreOptions) (string, error) {
	return formatRelativeTimeCorePrepared(value, formatter.data, options)
}

func (formatter RelativeTimeCoreFormatter) FormatToParts(value any, options RelativeTimeCoreOptions) ([]Part, error) {
	formatted, err := formatter.Format(value, options)
	if err != nil {
		return nil, err
	}
	return []Part{{"type": "text", "value": formatted}}, nil
}

func (formatter RelativeTimeCoreFormatter) FunctionRegistry() FunctionRegistry {
	return PortableFunctionRegistry().WithFunction("relativeTime", func(call FunctionCall) (string, error) {
		value := call.RawValue
		if value == nil {
			value = call.Value
		}
		style, err := callStringOption(call, "style", RelativeTimeCoreStyleShort)
		if err != nil {
			return "", err
		}
		numeric, err := callStringOption(call, "numeric", RelativeTimeCoreNumericAlways)
		if err != nil {
			return "", err
		}
		policy, err := callStringOption(call, "policy", RelativeTimeCorePolicyPrecise)
		if err != nil {
			return "", err
		}
		unit, err := callStringOption(call, "unit", RelativeTimeCoreUnitAuto)
		if err != nil {
			return "", err
		}
		options := RelativeTimeCoreOptions{
			Locale:  call.Locale,
			Style:   style,
			Numeric: numeric,
			Policy:  policy,
			Unit:    unit,
		}
		formatted, err := formatter.Format(value, options)
		if err == nil {
			return formatted, nil
		}
		code, ok := err.(Error)
		sourceValue, hasSourceValue := relativeTimeCoreSourceValue(call.InheritedSource)
		if !ok || code.Code != "bad-operand" || !hasSourceValue {
			return "", err
		}
		return formatter.Format(sourceValue, options)
	})
}

func relativeTimeCoreSourceValue(source *FunctionSource) (string, bool) {
	for current := source; current != nil; current = current.Inherited {
		if isDecimalSourceFunction(current.Function) {
			return current.Value, true
		}
	}
	return "", false
}

func formatRelativeTimeCorePrepared(value any, data relativeTimeCorePreparedData, options RelativeTimeCoreOptions) (string, error) {
	style, err := relativeTimeCoreOptionOneOf(defaultString(options.Style, RelativeTimeCoreStyleShort), "style", relativeTimeCoreStyleValues)
	if err != nil {
		return "", err
	}
	numeric, err := relativeTimeCoreOptionOneOf(defaultString(options.Numeric, RelativeTimeCoreNumericAlways), "numeric", relativeTimeCoreNumericValues)
	if err != nil {
		return "", err
	}
	policy, err := relativeTimeCoreOptionOneOf(defaultString(options.Policy, RelativeTimeCorePolicyPrecise), "policy", relativeTimeCorePolicyValues)
	if err != nil {
		return "", err
	}
	unit, err := relativeTimeCoreOptionOneOf(defaultString(options.Unit, RelativeTimeCoreUnitAuto), "unit", relativeTimeCoreUnitValues)
	if err != nil {
		return "", err
	}
	seconds, err := parseRelativeTimeCoreFinite(value)
	if err != nil {
		return "", err
	}
	selectedUnit := unit
	if unit == RelativeTimeCoreUnitAuto {
		selectedUnit = selectRelativeTimeCoreUnit(seconds, policy)
	}
	quantity, err := relativeTimeCoreQuantity(seconds, selectedUnit)
	if err != nil {
		return "", err
	}
	locale, err := localeOption(options.Locale, defaultRelativeTimeCoreLocale)
	if err != nil {
		return "", err
	}

	if useRelativeTimeCoreZero(policy, numeric, seconds) {
		if relative, ok, err := relativeTimeCoreTerm(data, locale, style, selectedUnit, "0"); err != nil {
			return "", err
		} else if ok {
			return relative, nil
		}
	}
	if numeric == RelativeTimeCoreNumericAuto {
		if offset, ok := relativeTimeCoreOffset(seconds, selectedUnit, quantity); ok {
			if relative, ok, err := relativeTimeCoreTerm(data, locale, style, selectedUnit, offset); err != nil {
				return "", err
			} else if ok {
				return relative, nil
			}
		}
	}

	direction := "future"
	if isNegativeRelativeTimeCore(seconds) {
		direction = "past"
	}
	category := selectCardinal(locale, quantity)
	pattern, err := relativeTimeCorePattern(data, locale, style, selectedUnit, direction, category)
	if err != nil {
		return "", err
	}
	return strings.ReplaceAll(pattern, "{0}", strconv.Itoa(quantity)), nil
}

func prepareRelativeTimeCoreData(data RelativeTimeCoreData) (relativeTimeCorePreparedData, error) {
	if len(data.LocaleMap) == 0 || len(data.PatternSets) == 0 {
		return relativeTimeCorePreparedData{}, mf2Error("missing-locale-data", "Relative-time core data has an unsupported shape.")
	}
	patternSets := map[string]map[string]map[string]RelativeTimeCoreUnitData{}
	for _, item := range data.PatternSets {
		if item.ID != "" && len(item.Data) > 0 {
			patternSets[item.ID] = item.Data
		}
	}
	if len(patternSets) == 0 {
		return relativeTimeCorePreparedData{}, mf2Error("missing-locale-data", "Relative-time core data has an unsupported shape.")
	}
	return relativeTimeCorePreparedData{localeMap: data.LocaleMap, patternSets: patternSets}, nil
}

func parseRelativeTimeCoreFinite(value any) (float64, error) {
	if value == nil {
		return 0, badOperand("Relative-time core requires a finite numeric value.")
	}
	if _, ok := value.(bool); ok {
		return 0, badOperand("Relative-time core requires a finite numeric value.")
	}
	text := strings.TrimSpace(fmt.Sprint(value))
	if text == "" {
		return 0, badOperand("Relative-time core requires a finite numeric value.")
	}
	if runeCountExceeds(text, maxRelativeTimeCoreOperandLength) {
		return 0, badOperand("Relative-time core requires a finite numeric value.")
	}
	if !relativeTimeCoreDecimalNumberPattern.MatchString(text) {
		return 0, badOperand("Relative-time core requires a finite numeric value.")
	}
	parsed, err := strconv.ParseFloat(text, 64)
	if err != nil || math.IsInf(parsed, 0) || math.IsNaN(parsed) {
		return 0, badOperand("Relative-time core requires a finite numeric value.")
	}
	return parsed, nil
}

func relativeTimeCoreOptionOneOf(value, name string, allowed []string) (string, error) {
	if runeCountExceeds(value, maxRelativeTimeCoreOptionLength) {
		return "", badOption(name + " must not exceed 256 characters.")
	}
	for _, candidate := range allowed {
		if value == candidate {
			return value, nil
		}
	}
	return "", badOption(name + " must be one of " + strings.Join(allowed, ", ") + ".")
}

func selectRelativeTimeCoreUnit(seconds float64, policy string) string {
	absolute := math.Abs(seconds)
	for _, step := range relativeTimeCorePolicies[policy] {
		if absolute < step.upper {
			return step.unit
		}
	}
	return RelativeTimeCoreUnitYear
}

func relativeTimeCoreQuantity(seconds float64, unit string) (int, error) {
	absolute := math.Abs(seconds)
	if absolute == 0 {
		return 0, nil
	}
	quantity := math.Max(1, math.Floor(absolute/relativeTimeCoreUnitSeconds[unit]+0.5))
	if quantity > maxRelativeTimeCoreQuantity {
		return 0, badOperand("Relative-time core quantity is outside the supported range.")
	}
	return int(quantity), nil
}

func useRelativeTimeCoreZero(policy, numeric string, seconds float64) bool {
	return policy == RelativeTimeCorePolicyChat && numeric == RelativeTimeCoreNumericAuto && math.Abs(seconds) < 45
}

func relativeTimeCoreOffset(seconds float64, unit string, quantity int) (string, bool) {
	if quantity == 0 {
		return "0", true
	}
	if math.Abs(seconds) != float64(quantity)*relativeTimeCoreUnitSeconds[unit] {
		return "", false
	}
	if isNegativeRelativeTimeCore(seconds) {
		return "-" + strconv.Itoa(quantity), true
	}
	return strconv.Itoa(quantity), true
}

func isNegativeRelativeTimeCore(seconds float64) bool {
	return math.Signbit(seconds)
}

func relativeTimeCoreTerm(
	data relativeTimeCorePreparedData,
	locale string,
	style string,
	unit string,
	offset string,
) (string, bool, error) {
	unitData, err := relativeTimeCoreUnitData(data, locale, style, unit)
	if err != nil {
		return "", false, err
	}
	relative, ok := unitData.Relative[offset]
	return relative, ok, nil
}

func relativeTimeCorePattern(
	data relativeTimeCorePreparedData,
	locale string,
	style string,
	unit string,
	direction string,
	category string,
) (string, error) {
	unitData, err := relativeTimeCoreUnitData(data, locale, style, unit)
	if err != nil {
		return "", err
	}
	patterns := unitData.Future
	if direction == "past" {
		patterns = unitData.Past
	}
	if pattern, ok := patterns[category]; ok {
		return pattern, nil
	}
	if pattern, ok := patterns["other"]; ok {
		return pattern, nil
	}
	return "", mf2Error("missing-locale-data", "Missing relative-time pattern for "+locale+"/"+style+"/"+unit+"/"+direction+".")
}

func relativeTimeCoreUnitData(
	data relativeTimeCorePreparedData,
	locale string,
	style string,
	unit string,
) (RelativeTimeCoreUnitData, error) {
	patternSet, err := relativeTimeCorePatternSetFor(data, locale)
	if err != nil {
		return RelativeTimeCoreUnitData{}, err
	}
	styleData, ok := patternSet[style]
	if !ok {
		return RelativeTimeCoreUnitData{}, mf2Error("missing-locale-data", "Missing relative-time unit data for "+locale+"/"+style+"/"+unit+".")
	}
	unitData, ok := styleData[unit]
	if !ok {
		return RelativeTimeCoreUnitData{}, mf2Error("missing-locale-data", "Missing relative-time unit data for "+locale+"/"+style+"/"+unit+".")
	}
	return unitData, nil
}

func relativeTimeCorePatternSetFor(
	data relativeTimeCorePreparedData,
	locale string,
) (map[string]map[string]RelativeTimeCoreUnitData, error) {
	for _, candidate := range localeLookupChain(locale) {
		setID, ok := data.localeMap[candidate]
		if !ok {
			continue
		}
		patternSet, ok := data.patternSets[setID]
		if !ok {
			return nil, mf2Error("missing-locale-data", "Missing relative-time pattern set "+setID+".")
		}
		return patternSet, nil
	}
	return nil, mf2Error("missing-locale-data", "Missing relative-time locale data for "+locale+".")
}
